### ISSUE TITLE
Lab WP loses its number

### ISSUE THREAD

**Issue by bekuno** [cgeo/cgeo#15210]
### Describe your problem!

I have a lab with ordered WP.
In the map view the WP have their correct number.
But when only showing the WP, then he has the number 0.

See here.
Map view:
![Screenshot_20240115_085458_cgeo 3](https://github.com/cgeo/cgeo/assets/32555742/ebe42229-454e-44e3-94f4-d8c6c42bfa1b)

Single WP view:
![Screenshot_20240115_085731_cgeo 3](https://github.com/cgeo/cgeo/assets/32555742/471a9614-7f34-4704-ae14-d5391668c1ad)

### How to reproduce?

- Have a ordered Lab visible on map
- tap a WP of this lab
- In the WP overview, tap the compass symbol
- In the compass menu, tap the map symbol

### Actual result after these steps?

the WP has number 0

### Expected result after these steps?

The WP has his correct number.
If this is not possible, then the WP should have no number.

### Reproducible

Yes

### c:geo Version

2024.01.06

**Comment by ztNFny** [comment]
I'm unable to reproduce this on latest master. Is there some other condition that has to be met in order to repoduce?

Could you put a breakpoint in MapMarkerUtils:294 (stageCounter = Integer.parseInt(waypoint.getPrefix());) to see what is passed as "waypoint" resp waypoint.getPrefix()?

I tested:
1. Store linear lab
2. View list as map / View cache as map
3. Tap WP
4. In the WP overview, tap the compass symbol
5. In the compass menu, tap the map symbol

**Comment by moving-bits** [comment]
Different behavior, but may be related:
- I've opened Berlin on map, set a filter to view only labs, and enabled live mode
- downloaded a couple of labs
- (stages are not showing automatically on successful download, but that's a separate issue)
- closed the map and reopened it
- now the stages show up as a "puzzle piece" icon
- I open the related cache and swipe to the waypoint page => here the waypoint icons are numbered correctly
- I force-closed c:geo by means of "close all apps" and reopened the map at the same location
- waypoints still appear as "puzzle piece" icons only, same is in info sheet, and only on opening the cache + swiping to waypoints page, they show up as numbered icons

**Comment by ztNFny** [comment]
> * waypoints still appear as "puzzle piece" icons only, same is in info sheet

Sounds like that at that stage the cache is either not loaded at all or incomplete. The condition for showing the numbers is:
`cache != null && cache.isLinearAlc()`

Does unified map load WPs without their cache?

**Comment by moving-bits** [comment]
> Does unified map load WPs without their cache?

This can happen, depending on how the map is being called. I would need to look up in code, in which order caches and waypoints are being loaded in that specific case (plain "show map").

Quick cross-check with "old" NewMap seems to not have this issue I've described above. (But only have smoke-tested this.)

**Comment by ztNFny** [comment]
If that's the case there are 2 approaches:

1. Load caches before WPs
2. Store the info (linear) with the WP instead of with the cache

Imho 1 would be nicer, but if it's too complicated 2 shouldn't be too hard I guess.

**Comment by moving-bits** [comment]
For UnifiedMap in "plain map" mode, it seems to be the `LoadInBackgroundHandler` organizing the loading of caches and waypoints. Lines 94-96 are:

```
final SearchResult searchResult = new SearchResult(DataStore.loadCachedInViewport(viewport.resize(1.2)));
...
final Set<Geocache> cachesFromSearchResult = searchResult.getCachesFromSearchResult(LoadFlags.LOAD_WAYPOINTS);
```

That should load caches first, and waypoints only then.

BUT: Another observation: If I switch off "live mode", waypoint icons show their number!
So it might be an outdated `SearchResult` for live mode overriding the already existing information...

**Comment by ztNFny** [comment]
> If that's the case there are 2 approaches:
> 
> 1. Load caches before WPs
> 2. Store the info (linear) with the WP instead of with the cache

Thinking about it a bit more: 2 is not an option as other data is loaded from cache as well, e.g..custom markers, visited state, ...

**Comment by moving-bits** [comment]
As for the original issue mentioned in opening post:

> I'm unable to reproduce this on latest master. Is there some other condition that has to be met in order to repoduce?

I can easily reproduce it on both current release 2024.01.06 and current state of `release` branch. The key for me was to use `NewMap` or `CGeoMap` implementations, not `UnifiedMap` - with `UnifiedMap` this original issue seems to work.

**Comment by moving-bits** [comment]
Thus as we have two different issues here, I'll create a separate issue for yesterday's finding.

WRT to the original issue, we need to decide whether it should be fixed or not. I'll put it on the respective project list.

**Comment by ztNFny** [comment]
I was unable to reproduce with NewMap (not unified) as well. If you can reliably reproduce debugging the mentioned line would be great.

**Comment by moving-bits** [comment]
Cache is available, and also `alcMode == 1` returns true, so it continues with the block for creating a number overlay.

But: The field "waypoint prefix" is empty, thus parsing it throws an exception, and `stageCounter` defaults to 0.

**Comment by ztNFny** [comment]
> Cache is available, and also `alcMode == 1` returns true, so it continues with the block for creating a number overlay.
> 
> But: The field "waypoint prefix" is empty, thus parsing it throws an exception, and `stageCounter` defaults to 0.

Thanks for debugging this. So no issue in MapMarkerUtils but somewhere between the CachePopupActivity and MapActivity handovers the prefix is lost, probably incomplete objects are handed over. 
If the issue is Unreproducable on Unified Map I'd question whether it's worth fixing as those handovers are probably going to be a complicated thing.

**Comment by moving-bits** [comment]
That rings a bell: Isn't it that the "old" maps use just the waypoint's coordinates on setting a waypoint as target? In contrast to this, UnifiedMap uses the actual waypoint object, so all properties are available here.

**Comment by moving-bits** [comment]
Fix prepared with #15327
